src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs

```diff
@@ -0,0 +1,729 @@
+        /// <returns>The list of IDs with special tokens added.</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="tokenIds0"/> is null.</exception>
+        public IReadOnlyList<int> BuildInputsWithSpecialTokens(IEnumerable<int> tokenIds0, IEnumerable<int>? tokenIds1 = null)
+        {
+            if (tokenIds0 is null)
+            {
+                throw new ArgumentNullException(nameof(tokenIds0));
+            }
+
+            // Add 2 for [CLS] and [SEP] tokens. Add 1 for [SEP] token if tokenIds1 is not null.
+            int capacity = tokenIds0.Count() + 2 + (tokenIds1 is null ? 0 : tokenIds1.Count() + 1);
+            List<int> ids = new List<int>(capacity: capacity) { ClsTokenId };
+            ids.AddRange(tokenIds0);
+            ids.Add(SepTokenId);
+
+            if (tokenIds1 is not null)
+            {
+                ids.AddRange(tokenIds1);
+                ids.Add(SepTokenId);
+            }
+
+            return ids;
+        }
+
+        /// <summary>
+        /// Build model inputs from a sequence or a pair of sequence for sequence classification tasks by concatenating and adding special tokens. A BERT sequence has the following format:
+        ///     - single sequence: `[CLS] tokenIds0 [SEP]`
+        ///     - pair of sequences: `[CLS] tokenIds0 [SEP] tokenIds1 [SEP]`
+        /// </summary>
+        /// <param name="tokenIds0">List of IDs to which the special tokens will be added.</param>
+        /// <param name="buffer">The buffer to write the token IDs with special tokens added.</param>
+        /// <param name="written">The number of elements written to the buffer.</param>
+        /// <param name="tokenIds1">Optional second list of IDs for sequence pairs.</param>
+        /// <returns>The status of the operation.</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="tokenIds0"/> is null.</exception>
+        public OperationStatus BuildInputsWithSpecialTokens(IEnumerable<int> tokenIds0, Span<int> buffer, out int written, IEnumerable<int>? tokenIds1 = null)
+        {
+            if (tokenIds0 is null)
+            {
+                throw new ArgumentNullException(nameof(tokenIds0));
+            }
+
+            // Add 2 for [CLS] and [SEP] tokens. Add 1 for [SEP] token if tokenIds1 is not null.
+            int capacity = tokenIds0.Count() + 2 + (tokenIds1 is null ? 0 : tokenIds1.Count() + 1);
+            if (buffer.Length < capacity)
+            {
+                written = 0;
+                return OperationStatus.DestinationTooSmall;
+            }
+
+            written = 0;
+            buffer[written++] = ClsTokenId;
+            foreach (int id in tokenIds0)
+            {
+                buffer[written++] = id;
+            }
+            buffer[written++] = SepTokenId;
+
+            if (tokenIds1 is not null)
+            {
+                foreach (int id in tokenIds1)
+                {
+                    buffer[written++] = id;
```

Comment:
For example, if this second enumeration of tokenIds1 produces fewer elements than in the capacity computation above, this will likely result in an IndexOutOfRangeException.